core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -0,0 +1,951 @@
+         * Using VALUE_LENGTH_MASK and VALUE_BLOCK_SHIFT the blockID and length can be extracted.
+         * Currently, the length of a value is limited to 8MB, and blockID is limited to 512 blocks
+         * (with the current block size of 256MB, the total memory is up to 128GB).
+         *
+         * VALUE_BLOCK
+         *
+         * VALUE_LENGTH
+         *
+         *
+         * VALUE_VERSION - as the name suggests this is the version of the value reference by VALUE_REFERENCE.
+         * It initially equals to INVALID_VERSION.
+         * If an entry with version v is removed, then this field is CASed to be -v after the value is marked
+         * off-heap and the value reference becomes INVALID_VALUE.
+         */
+        NEXT(0), KEY_REFERENCE(1), KEY_POSITION(1), KEY_BLOCK_AND_LENGTH(2), KEY_BLOCK(2), KEY_LENGTH(2),
+        VALUE_REFERENCE(3), VALUE_POSITION(3), VALUE_BLOCK_AND_LENGTH(4), VALUE_BLOCK(4), VALUE_LENGTH(4),
+        VALUE_VERSION(5);
+
+        final int value;
+
+        OFFSET(int value) {
+            this.value = value;
+        }
+    }
+
+    static final long   INVALID_VALUE_REFERENCE = 0;
+    static final long   INVALID_KEY_REFERENCE = 0;
+    private static final int INVALID_ENTRY_INDEX = 0;
+    private static final int    BLOCK_ID_LENGTH_ARRAY_INDEX = 1;
+    private static final int    POSITION_ARRAY_INDEX = 0;
+
+    // location of the first (head) node - just a next pointer (always same value 0)
+    private final int headNextIndex;
+
+    // index of first item in array, after head (not necessarily first in list!)
+    private static final int FIRST_ITEM = 1;
+
+    private static final int FIELDS = 6;  // # of fields in each item of entries array
+    // key block is part of key length integer, thus key length is limited to 65KB
+    private static final int KEY_LENGTH_MASK = 0xffff; // 16 lower bits
+    private static final int KEY_BLOCK_SHIFT = 16;
+    // Assume the length of a value is up to 8MB because there can be up to 512 blocks
+    private static final int VALUE_LENGTH_MASK = 0x7fffff;
+    private static final int VALUE_BLOCK_SHIFT = 23;
+
+    private static final Unsafe unsafe = UnsafeUtils.unsafe;
+    private final MemoryManager memoryManager;
+
+    private final int[] entries;    // array is initialized to 0, i.e., Chunk.NONE_NEXT - this is important!
+    private final int entriesCapacity; // number of entries (not ints) to be maximally held
+
+    private final AtomicInteger nextFreeIndex;    // points to next free index of entry array
+    private final AtomicInteger numOfEntries;     // counts number of entries inserted & not deleted
+    // for writing the keys into the off-heap bytebuffers (Blocks)
+    private final OakSerializer<K> keySerializer;
+    private final OakSerializer<V> valueSerializer;
+
+    private final ValueUtils valOffHeapOperator; // is used for any value off-heap metadata access
```

Comment:
Utils classes are usually static. Why do you need an attribute for the util? are there more than one util classes?

Reply:
We can deal with it in different PR